protocols/tls13/src/tls13.rs

```diff
@@ -0,0 +1,213 @@
+#![allow(dead_code)]
+#![allow(unused_variables)]
+#![allow(non_camel_case_types)]
+#![allow(non_snake_case)]
+#![allow(non_upper_case_globals)]
+#![allow(unused_imports)]
+#![allow(unused_parens)]
+
+mod tls13formats;
+use tls13formats::*;
+mod cryptolib;
+use cryptolib::*;
+mod tls13record;
+use tls13record::*;
+mod tls13handshake;
+use tls13handshake::*;
+mod testtraces;
+use testtraces::*;
+
+// Import hacspec and all needed definitions.
+use hacspec_lib::*;
+
+// Client-Side API for TLS 1.3 Applications
+// client_init -> (client_send0)* -> client_finish -> (client_send1|client_recv1)* */
+
+pub struct Client0(ALGS,TranscriptClientHello,ClientPostClientHello,Option<(CipherState,KEY)>);
+pub struct Client1(ALGS,TranscriptClientFinished,ClientPostClientFinished,CipherState,CipherState,KEY);
+    
+pub fn client_init(algs:ALGS,sn:&Bytes,tkt:Option<Bytes>,psk:Option<KEY>,ent:Entropy) -> Res<(Bytes,Client0)> {
+    let ALGS(ha, ae, sa, gn, psk_mode, zero_rtt) = &algs;
+    let (crand,gx,cstate) = get_client_hello(algs,psk,ent)?;
+    let (ch,trunc_len) = client_hello(&algs,&crand,&gx,sn,&tkt)?;
+    let tx_trunc = transcript_truncated_client_hello(algs,&ch,trunc_len)?;
+    let binder = get_client_hello_binder(&tx_trunc,&cstate)?;
+    let nch = set_client_hello_binder(&algs,&binder,ch)?;
+    let tx_ch = transcript_client_hello(algs,&nch)?;
+    let rec = handshake_record(&nch)?;
+    let c2s0 = client_get_0rtt_keys(&tx_ch,&cstate)?;
+    let st = Client0(algs,tx_ch,cstate,c2s0);
+    Ok((rec,st))  
+}
+
+pub fn client_send0(st:Client0,msg:&Bytes) -> Res<(Bytes,Client0)> {
+    let Client0(algs,transcript,cstate,c2s0) = st;
+    if let Some((c2s,exp)) = c2s0 {
+        let (cip,c2s) = encrypt_record(ct_app_data,msg,0,c2s)?;
+        Ok((cip,Client0(algs,transcript,cstate,Some((c2s,exp)))))
+    } else {Err(psk_mode_mismatch)}
+}
+
+pub fn client_finish(msg:&Bytes,st:Client0) -> Res<(Bytes,Client1)> {
+    let Client0(algs,tx_ch,cstate,_) = st;
+    let ALGS(ha, ae, sa, gn, psk_mode, zero_rtt) = &algs;
+    let (sh,len1) = check_handshake_record(&msg)?;
+    let (sr,gy) = parse_server_hello(&algs,&sh)?;    
+    let tx_sh = transcript_server_hello(tx_ch,&sh)?;
+    let (c2s,s2c,cstate) = put_server_hello(sr, gy, algs, &tx_sh, cstate)?;
+    let len2 = check_encrypted_record(&msg.slice_range(len1..msg.len()))?;
+    let (ct,payload,s2c) = decrypt_record(&msg.slice_range(len1..len1+len2),s2c)?;
+    if ct == ct_handshake {
+        let mut next = 0;
+        let (ee,len3) = check_handshake_message(&payload)?;
+        parse_encrypted_extensions(&algs,&ee)?;
+        next = next + len3;
+        let (tx_cv,cstate) =
```

Comment:
So this mixture of expressions and statements is going to be hard for me to deal with semantically. I would prefer something like

```
let mut tx_cv = ...;
let mut cstate = ...;
if psk_mode { ... tx_cv = ...; cstate = ... } else { ... tx_cv = ...; cstate = ... }
```

Reply:
This can be done, but is there a "default" value I could give to these? They have very specific types, and I'd rather not go the Option route, but I could.

Reply:
Didn't we say we want to allow `if let`?